falcon/app_helpers.py

```diff
@@ -183,7 +184,8 @@ def default_serialize_error(req, resp, exception):
 
     if preferred is not None:
         if preferred == 'application/json':
-            resp.body = exception.to_json()
+            dumps = getattr(resp.options.media_handlers.get(MEDIA_JSON), 'dumps', None)
```

Comment:
This is a fair bit of dark magic that IMHO would benefit from a docstring or a `NOTE` explaining how this works in different scenarios: `MEDIA_JSON` is not in `media_handlers`, `MEDIA_JSON` is in media handlers, but it is a custom handler not employing `dumps` etc.

Furthermore, this can potentially lead to hard-to-debug scenarios where the handler does define `dumps`, but the signature is not as we expect here, for instance, it requires an extra indentation level parameter, or what not. We should at minimum discuss this (the special meaning of `dumps`) in the custom handlers docs.

Alternatively, we could be checking the handler type, i.e. if it is strictly of type `JSONHandler`, or a subclass of it (?).

Reply:
I think we could just document this in media, specifying that loads and dumps of the json handler if exits and has that attributes are used by the application to handler (de)serialization of json.

Reply:
> but the signature is not as we expect here

There are at least three edge cases to address, and to ensure are documented for the developer:

1. A JSON handler is not defined at all. 
2. A JSON handler is defined using ``JSONHandler`` but the dumps/loads require parameters. (We should be OK on this one already since the docs talk about it and recommend ``functools.partial``).
3. A JSON handler is defined but it does not derive from ``JSONHandler``.  In this case it is possible that dumps/loads attributes are defined but not as expected. (The code already falls back to stdlib if dumps/loads are not found... alternatively would it make sense to use the `serialize()` method, although that would require extra branches in the logic since the return type is different?)

TBH I think it would be a more prudent design in the abstract to stick with the ``BaseHandler`` serialize/deserialize interface methods if it doesn't make things too convoluted. That would prevent us from coupling too tightly with the specific ``JSONHandler`` implementation.

Reply:
In the end I've opted to use the handler directly